Gmail Node

erin2722 opened the pull request:
:sparkles: here is the Gmail node.

In addition to code review, I would love to get feedback on:
-  📧  Any additional functionality you would like to see in it.
- ⚡ The ease of use. Especially for the labels and the replying functions. The data Gmail needs for API requests (message id, thread id) creates a very weird flow for making certain functionality work, so any advice on that would be much appreciated.

ivov commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,242 @@
+				} else if(['add', 'delete'].includes(operation)) {
+					// https://developers.google.com/gmail/api/v1/reference/users/messages/modify
+					const messageID = this.getNodeParameter('messageID', i);
+					const labelID = this.getNodeParameter('labelID', 0);
```
I have not checked the logic yet, but `0` instead of `i` _might_ not be right here. Please double check.

ivov commented on packages/nodes-base/nodes/Google/Gmail/EmailDescriptions.ts:
```diff
@@ -0,0 +1,251 @@
+				]
+			},
+		},
+		placeholder: '<CAHNQoFsC6JMMbOBJgtjsqN0eEc+gDg2a=SQj-tWUebQeHMDgqQ@mail.gmail.com>',
```
I haven't tested the node yet and don't know much about Message-IDs, but this placeholder seems to suggest that I should input `<` and `>` as well. Just commenting to make sure this is correct.

ivov commented on packages/nodes-base/nodes/Google/Gmail/GenericFunctions.ts:
```diff
@@ -0,0 +1,126 @@
+	}
+}
+
+export function encodeEmail(email: any = {}) { // I feel like I should not have to use any here
```
After the colon you should describe the shape of the object that you'll pass into `encodeEmail()`.

This is just an arbitrary example without looking at your code in detail:

```ts
type myEmail = {
	to: string; // a property called `to` of type `string` _must_ be present
	from: string;
	mimeVersion: number;
	cc?: Array<string>; // the property `cc` of type `Array<string>` may or may not be present
}
```
And then:

```ts
export function encodeEmail(email: myEmail) {
    // ...
}
```

Or you can insert the whole description after the colon, without naming the type. Interfaces are used similarly.

This way, if you try to pass in an object without the required properties, or with the required properties but with wrong types, TypeScript will let you know before you even run your code.

erin2722 replied:
@ivov Thanks for the initial feedback!! I just renamed according to conventions, got rid of the <> around message-id for easier use, and replaced `0` with `i` in a few places. I also created interfaces to get rid of the TS errors.

erin2722 commented:
@ivov Thanks for the initial feedback!! I just renamed according to conventions, got rid of the <> around message-id for easier use, and replaced 0 with i in a few places. I also created interfaces to get rid of the TS errors.

ivov commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,254 @@
+	labelFields,
+} from './LabelDescription';
+
+import * as _ from 'lodash';
```
If you're only using `isEmpty`, perhaps it'd be better to only `import { isEmpty } from 'lodash';`?

erin2722 replied:
Haha to be honest I used this code from Ricardo's Mandrill code, and you totally right that I should just use the built-in map. Just pushed a fix!

ivov commented on packages/nodes-base/nodes/Google/Gmail/LabelDescription.ts:
```diff
@@ -0,0 +1,105 @@
+			{
+				name: 'Add',
+				value: 'add',
+				description: 'Add a label.'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/LabelDescription.ts:
```diff
@@ -0,0 +1,105 @@
+				],
+				operation: [
+					'add',
+					'delete'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/LabelDescription.ts:
```diff
@@ -0,0 +1,105 @@
+				operation: [
+					'add',
+					'delete',
+					'get'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/LabelDescription.ts:
```diff
@@ -0,0 +1,105 @@
+		displayOptions: {
+			show: {
+				resource: [
+					'label'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+		displayOptions: {
+			show: {
+				resource: [
+					'message'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+				],
+				operation: [
+					'reply',
+					'create'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+				],
+				operation: [
+					'reply',
+					'create'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+		displayOptions: {
+			show: {
+				resource: [
+					'message'
```
Trailing comma

ivov commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+				],
+				operation: [
+					'create',
+					'reply'
```
Trailing comma

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/DraftDescription.ts:
```diff
@@ -0,0 +1,209 @@
+		options: [
+			{
+				displayName: 'To Email',
+				name: 'to_list',
```
variables should follow camelcasing conventions

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/DraftDescription.ts:
```diff
@@ -0,0 +1,209 @@
+			},
+			{
+				displayName: 'CC Email',
+				name: 'cc_list',
```
camel case

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/DraftDescription.ts:
```diff
@@ -0,0 +1,209 @@
+			},
+			{
+				displayName: 'BCC Email',
+				name: 'bcc_list',
```
same as above

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/GenericFunctions.ts:
```diff
@@ -0,0 +1,125 @@
+			"In-Reply-To: ", email.inReplyTo, "\n",
+			"References: ", email.reference, "\n",
+			"Subject: ", email.subject, "\n\n",
+			email.body
```
trailing comma

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,395 @@
+					body = {
+						'labelListVisibility': 'labelShow',
+						'messageListVisibility': 'show',
+						'name': labelName
```
comma

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,395 @@
+
+						body = {
+							'message': {
+							'raw': encodedMessage
```
comma

Shraddha2104 commented on packages/nodes-base/nodes/Google/Gmail/MessageDescription.ts:
```diff
@@ -0,0 +1,267 @@
+	},
+	{
+        displayName: 'To Email',
+        name: 'to_list',
```
camelcase

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,395 @@
+				//-------------------------------
+				// Add/Remove Label
+				//-------------------------------
+				} else if(['add', 'delete'].includes(operation)) {
```
I think generally this is how it is done in the codebase-- takes up less space

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,395 @@
+								// @ts-ignore
+								attachmentsBinary = attachmentsUi.attachmentsBinary.map((o) => {
+									if (items[i].binary!.hasOwnProperty(o.property)) {
+										const aux: Attachments = {name: '', content: '', type: ''};
+										aux.name = items[i].binary![o.property].fileName || 'unknown';
+										aux.content = items[i].binary![o.property].data;
+										aux.type = items[i].binary![o.property].mimeType;
+										return aux;
```
I think I like the explicit typing of Attachments here, but I will ask Ricardo

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,395 @@
+					//------------------------------------
+					// Reply to a message
+					//------------------------------------
+					} else if(operation === 'reply' && resource === 'message') {
```
Generally yes, but for this flow, I think that processing messages and drafts together makes a lot of sense bc they are so similar. It could totally be wrong, so I will ask Ricardo.

ivov replied:
Ah no, I meant simply `resource === 'message' && operation === 'reply'`

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/GenericFunctions.ts:
```diff
@@ -0,0 +1,125 @@
+	}
+
+
+    const encodedMail = new Buffer(str).toString("base64").replace(/\+/g, '-').replace(/\//g, '_');
+    return encodedMail;
```
I do not know regex too well but removing the `+` causes a lot of syntax errors haha. And totally right with the first comment.

ivov commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,386 @@
+
+					if(operation === 'add') {
+						body = {
+							'addLabelIds': [
```
Super minor detail, but no need for quotes in this kind of properties. Same for below.

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,386 @@
+		const operation = this.getNodeParameter('operation', 0) as string;
+
+		let method = '';
+		let body: IDataObject = {};
+		let qs: IDataObject = {};
```
I totally agree and was implementing it that way but learned that if a property is nested within another then TS does not allow you to set it. `body.message.raw` causes an `Object is possibly 'null' or 'undefined'` error. 😞

ivov replied:
It might be worth checking if the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) fixes that.

erin2722 replied:
Oh wow I did not know about this!

erin2722 replied:
Unfortunately that gives me this error `The left-hand side of an assignment expression may not be an optional property access.`

erin2722 commented on packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts:
```diff
@@ -0,0 +1,386 @@
+		let body: IDataObject = {};
+		let qs: IDataObject = {};
+		let endpoint = '';
+		let responseData;
```
Not sure why, but a lot of other nodes are similar!
